[2424] Ajax trainee filtering

AbigailMcP opened the pull request:
### Context

https://trello.com/c/BPsUCM8i/2424-l-ajax-trainee-list-filtering

### Changes proposed in this pull request

Adds Ajax filtering to our existing filters on the trainee index page.

**Things to note:**

1. I've removed the `PaginatedFilter` and `Filter` component.
  - All the `PaginatedFilter` component did was pass variables through to the `Filter` component and include a `.scss` file
  - All the `Filter` component did was encapsulate some helpers and render html slots that were defined in the top `index.html.erb`.
All this felt needlessly complicated and the nesting made the Ajax implementation much harder (since there were no defined partials to render dynamically)

2. I've moved the couple of methods from `Filters::View` into a `FiltersHelper` module, and included the html in partials.

3. These partials are rendered via Ajax and targeted via their `js-` prefixed IDs.

### TODO
- ~How to update the page title (help!)~
- ~Think about accessibility (help!)~
- ~Jest tests - I will add!~

### Guidance to review
- Filter trainees on `/trainees`
- Check that:
- [ ] The correct trainees are returned
- [ ] The trainee count is correct
- [ ] The correct selected filters are displayed
- [ ] The url is updated
- [ ] The page title is updated
- [ ] The submit button is removed
- [ ] The trainee export is correct
- [ ] The pagination is correct, especially the counts
- [ ] The sorting retains the filters
- [ ] When you disable JS, everything works the way it did before
- [ ] Using VoiceOver (or equivalent), a user is notified that something has changed when the filters are updated
- [ ] The browser back button works 😬

AbigailMcP commented on app/helpers/filter_helper.rb:
```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
```
Copied directly from our `Filter` component, no new code.

AbigailMcP commented on app/views/trainees/_filters.html.erb:
```diff
@@ -0,0 +1,89 @@
+<form method="get" id="js-live-filter">
```
Pulled from `index.rb` - no new code except for addition of `js` ids to use as selectors.

AbigailMcP commented on app/views/trainees/_results.html.erb:
```diff
@@ -0,0 +1,57 @@
+<div class="app-records-actions">
```
Pulled from index.rb - no new code.

AbigailMcP commented:
![](https://github.trello.services/images/mini-trello-icon.png) [[L] Ajax trainee list filtering](https://trello.com/c/BPsUCM8i/2424-l-ajax-trainee-list-filtering)

defong commented on app/webpacker/scripts/live_search.js:
```diff
@@ -0,0 +1,61 @@
+      url: '/trainees',
+      dataType: 'json',
+      data: this.state
+    }).done(response => { this.displayResults(response) })
```
```suggestion
    }).done(this.displayResults)
```

AbigailMcP replied:
i do this every. single. time. will i ever learn?

AbigailMcP replied:
needs a `bind` though 😞

defong commented on app/webpacker/scripts/live_search.js:
```diff
@@ -0,0 +1,61 @@
+
+  formChange () {
+    this.saveState()
+    const pageUpdated = this.updatePage()
+    pageUpdated.done(() => { this.updateUrl() })
```
```suggestion
    this.updatePage().done(this.updateUrl)
```

AbigailMcP commented on app/views/trainees/index.html.erb:
```diff
@@ -14,140 +19,40 @@
-            <h2 class="govuk-heading-m">Records</h2>
+        <!-- Div made focusable to 'catch' focus from filters. see https://github.com/DFE-Digital/apply-for-teacher-training/pull/2640 -->
+        <div class="moj-filter__content" tabindex="-1">
+          <div id='js-selected-filters' aria-live="polite">
```
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions

AbigailMcP commented:
I've added an `aria-live=polite` to the selected filters div to announce to the user that the filters have changed, but I'm not sure what's best. Would it be nicer to announce the count? cc @edwardhorsford 

This is the selected filters announcement:

<img width="742" alt="Screenshot 2021-08-11 at 17 39 22" src="https://user-images.githubusercontent.com/18436946/129068902-39015328-a15b-4de7-afcf-8fe377c6314c.png">

ralph-hawkins commented:
> I've added an `aria-live=polite` to the selected filters div to announce to the user that the filters have changed, but I'm not sure what's best. Would it be nicer to announce the count? cc @edwardhorsford

I think its clearer if voice over announces the number of results here as way to give the user an indication of the update to the page

ralph-hawkins commented:
It looks like this introduces a bug where the link to close filters at mobile widths is missing, so a user can't escape the filter menu after they've opened it.

AbigailMcP commented:
> > I've added an `aria-live=polite` to the selected filters div to announce to the user that the filters have changed, but I'm not sure what's best. Would it be nicer to announce the count? cc @edwardhorsford
> 
> I think its clearer if voice over announces the number of results here as way to give the user an indication of the update to the page

I've changed the aria-live so that it announces the page heading 'Trainee records (4)'.

AbigailMcP commented:
> It looks like this introduces a bug where the link to close filters at mobile widths is hidden, so a user can't escape the filter menu after they've opened it.

Thank you! Fixed now.

darokel commented on app/webpacker/scripts/live_filter.spec.js:
```diff
@@ -0,0 +1,89 @@
+describe('LiveFilter', () => {
+  beforeEach(() => {
+    const $form = $(`<form method="get" id="js-live-filter">
+    <input type="submit" name="commit" value="Apply filters" id="js-submit">
```
```suggestion
      <input type="submit" name="commit" value="Apply filters" id="js-submit">
      <input type="checkbox" name="level[]" id="level-early_years" value="early_years">
      <input type="checkbox" name="level[]" id="level-primary" value="primary">
      <select name="subject" id="subject" class="govuk-select"><option value="All subjects">All subjects</option>
        <option value="ancient hebrew">Ancient hebrew</option>
        <option value="applied biology">Applied biology</option>
        <option value="applied chemistry">Applied chemistry</option>
      </select>
```

darokel commented on app/webpacker/scripts/live_filter.spec.js:
```diff
@@ -0,0 +1,89 @@
+import $ from 'jquery'
+import LiveFilter from './live_filter'
+
+describe('LiveFilter', () => {
```
❤️

jhenderson commented:
This is a cool feature. Nicely done @AbigailMcP

defong commented on app/views/trainees/_action_bar.html.erb:
```diff
@@ -0,0 +1,12 @@
+    </div>
+  <% end %>
+  <%= render SortLinks::View.new %>
+<% end %>
```
line ending

AbigailMcP replied:
okkkkk

ralph-hawkins reviewed (approved):
nice 👍🏻
